File: Linked List/doubly/addition/list.cpp
// Doubly Linked List - Addition (C++)
// Insert at head and tail, then print forward and backward.

#include <iostream>
using namespace std;

struct Node {
    int data;
    Node* prev;
    Node* next;
    Node(int value) : data(value), prev(nullptr), next(nullptr) {}
};

Node* insertAtHead(Node* head, int value) {
    Node* node = new Node(value);
    if (head != nullptr) {
        node->next = head;
        head->prev = node;
    }
    return node;
}

void insertAtTail(Node*& head, int value) {
    Node* node = new Node(value);
    if (head == nullptr) { head = node; return; }
    Node* cur = head;
    while (cur->next != nullptr) cur = cur->next;
    cur->next = node; node->prev = cur;
}

Node* getTail(Node* head) {
    Node* cur = head;
    while (cur != nullptr && cur->next != nullptr) cur = cur->next;
    return cur;
}

void printForward(Node* head) {
    Node* cur = head;
    while (cur != nullptr) { cout << cur->data << " "; cur = cur->next; }
    cout << "\n";
}

void printBackward(Node* tail) {
    Node* cur = tail;
    while (cur != nullptr) { cout << cur->data << " "; cur = cur->prev; }
    cout << "\n";
}

int main() {
    Node* head = nullptr;
    insertAtTail(head, 20);
    insertAtTail(head, 30);
    head = insertAtHead(head, 10);
    insertAtTail(head, 40);

    cout << "Forward: ";
    printForward(head);
    cout << "Backward: ";
    printBackward(getTail(head));

    // cleanup
    Node* cur = head;
    while (cur != nullptr) { Node* d = cur; cur = cur->next; delete d; }
    return 0;
}

File: Linked List/doubly_circular/addition/list.py
# Doubly Circular Linked List - Addition (Python)
# Insert at head and tail.

class Node:
    def __init__(self, value: int) -> None:
        self.data = value
        self.prev: "Node | None" = None
        self.next: "Node | None" = None


def insert_at_head(head: "Node | None", value: int) -> Node:
    node = Node(value)
    if head is None:
        node.next = node; node.prev = node
        return node
    tail = head.prev
    node.next = head; node.prev = tail
    tail.next = node; head.prev = node
    return node


def insert_at_tail(head: "Node | None", value: int) -> Node:
    node = Node(value)
    if head is None:
        node.next = node; node.prev = node
        return node
    tail = head.prev
    tail.next = node; node.prev = tail; node.next = head; head.prev = node
    return head


def print_forward(head: "Node | None") -> None:
    if head is None:
        print()
        return
    cur = head
    while True:
        print(cur.data, end=" ")
        cur = cur.next
        if cur is head:
            break
    print()


if __name__ == "__main__":
    head = None
    head = insert_at_tail(head, 20)
    head = insert_at_tail(head, 30)
    head = insert_at_head(head, 10)
    head = insert_at_tail(head, 40)
    print_forward(head)


